python/src/specialize.cc

```diff
@@ -0,0 +1,884 @@
+  if (!tma_desc_cpu_ptr_attr)
+    goto cleanup;
+  _fields_attr = PyUnicode_InternFromString("_fields");
+  if (!_fields_attr)
+    goto cleanup;
+  block_shape_attr = PyUnicode_InternFromString("block_shape");
+  if (!block_shape_attr)
+    goto cleanup;
+  layout_attr = PyUnicode_InternFromString("layout");
+  if (!layout_attr)
+    goto cleanup;
+  has_native_tensor_spec_attr =
+      PyUnicode_InternFromString("supports_native_tensor_specialization");
+  if (!has_native_tensor_spec_attr)
+    goto cleanup;
+  get_tensor_spec_attr =
+      PyUnicode_InternFromString("get_tensor_specialization");
+  if (!get_tensor_spec_attr)
+    goto cleanup;
+
+  return;
+
+cleanup:
+  if (!PyErr_Occurred()) {
+    PyErr_SetString(PyExc_RuntimeError,
+                    "failed to initialize __triton_specialize module objects");
+  }
+  return;
+}
+
+static void cleanup_interned_strings() {
+  Py_XDECREF(jit_function_cls);
+  jit_function_cls = nullptr;
+  Py_XDECREF(tensor_descriptor_cls);
+  tensor_descriptor_cls = nullptr;
+  Py_XDECREF(gluon_tensor_descriptor_cls);
+  gluon_tensor_descriptor_cls = nullptr;
+  Py_XDECREF(canonicalize_dtype_fn);
+  canonicalize_dtype_fn = nullptr;
+  Py_XDECREF(canonicalize_ptr_dtype_fn);
+  canonicalize_ptr_dtype_fn = nullptr;
+  Py_XDECREF(constexpr_cls);
+  constexpr_cls = nullptr;
+  Py_XDECREF(torch_tensor_cls);
+  torch_tensor_cls = nullptr;
+
+  Py_XDECREF(i32_str);
+  i32_str = nullptr;
+  Py_XDECREF(i64_str);
+  i64_str = nullptr;
+  Py_XDECREF(u64_str);
+  u64_str = nullptr;
+  Py_XDECREF(fp32_str);
+  fp32_str = nullptr;
+  Py_XDECREF(u1_str);
+  u1_str = nullptr;
+  Py_XDECREF(D_str);
+  D_str = nullptr;
+  Py_XDECREF(constexpr_str);
+  constexpr_str = nullptr;
+  Py_XDECREF(empty_str);
+  empty_str = nullptr;
+  Py_XDECREF(nvTmaDesc_str);
+  nvTmaDesc_str = nullptr;
+  Py_XDECREF(data_ptr_attr);
+  data_ptr_attr = nullptr;
+  Py_XDECREF(dtype_attr);
+  dtype_attr = nullptr;
+  Py_XDECREF(cache_key_attr);
+  cache_key_attr = nullptr;
+  Py_XDECREF(tma_desc_cpu_ptr_attr);
+  tma_desc_cpu_ptr_attr = nullptr;
+  Py_XDECREF(_fields_attr);
+  _fields_attr = nullptr;
+  Py_XDECREF(base_attr);
+  base_attr = nullptr;
+  Py_XDECREF(block_shape_attr);
+  block_shape_attr = nullptr;
+  Py_XDECREF(layout_attr);
+  layout_attr = nullptr;
+  Py_XDECREF(has_native_tensor_spec_attr);
+  has_native_tensor_spec_attr = nullptr;
+  Py_XDECREF(get_tensor_spec_attr);
+  get_tensor_spec_attr = nullptr;
+}
+
+static void cleanup_caches() {
+  for (auto &[_, v] : dtype_ptr2str) {
+    Py_XDECREF(v);
+  }
+  dtype_ptr2str.clear();
+  for (auto &[_, v] : dtype2str) {
+    Py_XDECREF(v);
+  }
+  dtype2str.clear();
+  for (auto &[_, v] : tensor_desc_cache) {
+    Py_XDECREF(v);
+  }
+  tensor_desc_cache.clear();
+  type_handler_cache.clear();
+}
+
+static void cleanup_globals() {
+  // only cleanup if we are the same process that initialized the module
+  if (init_pid != getpid()) {
+    return;
+  }
```

Comment:
Where does this issue come up? 

tbh I think the best solution might just be to not register an atexit function at all. The assumption being that these live until the process is about to end anyway, so there is no point in trying to drop the references.